### PR TITLE
LoadEventNexus clean-up

### PR DESCRIPTION
**Pull request by peterfpeterson**
This was a `valgrind --tool=callgrind` guided optimization of `LoadEventNexus`. Unfortunately, it didn't create much of a net effect other than cleaning up the code some. The exact script used is
```python
from mantid.simpleapi import mtd, LoadEventNexus, FilterByTime, RebinToWorkspace
from collections import defaultdict 
import numpy as np
from os import path
import time

filenames = ['CNCS_7860_event.nxs', 
             '/SNS/EQSANS/IPTS-17292/0/68200/NeXus/EQSANS_68200_event.nxs', # 334MiB
             '/SNS/NOM/IPTS-23742/nexus/NOM_131564.nxs.h5', # 5.9GiB
             '/SNS/REF_L/IPTS-22484/nexus/REF_L_172383.nxs.h5'] 

results = defaultdict(lambda: list())
NUM_TRIES = 5
for filename in filenames:
    for i in range(NUM_TRIES):
        shortname = path.split(filename)[-1]
        tzero = time.time()
        print('ITERATION {}: {}'.format(i, filename))
        LoadEventNexus(Filename=filename, OutputWorkspace='wksp', LoadLogs=False)
        total_events=mtd['wksp'].getNumberEvents()
        print('total events: {}'.format(total_events))
        results[shortname].append(time.time() - tzero)

print('Results from {} tries'.format(NUM_TRIES))
for name, values in results.items():
    values = np.sort(values)[:-1] # drop slowest which is likely from cold cache
    print('{}: {:.1f} +/- {:.1f} seconds'.format(name, np.mean(values), np.std(values)))
```
with most of the development changes being driven by checks with the EQSANS file followed by the NOMAD file.

Results from 5 tries of `master` and 10 tries of new code (dropping slowest, times are in seconds)

| run        | size | num banks |  total events | `master` | branch |
|------------|------|---------------|-----------------|------------|----------|
| CNCS_7860       |  8.9MiB  | 44 | 112,266 | 1.3 +/- 0.0 | 1.4 +/- 0.0 |
| EQSANS_68200 | 334MiB | 48 | 76,862,885 | 9.8 +/- 0.1 | 9.6 +/- 0.2 |
| NOM_131564           | 5.9GiB   | 99 | 741,688,197 | 22.0 +/- 0.3 | 18.9 +/- 0.2 |
| REF_L_172383        | 21MiB    |   1 | 71,999 | 0.4 +/- 0.0 | 0.3 +/- 0.0 |\\

Things that were not properly explored were
1. Not using the threadpool for `LoadBankFromDisk`. Instead just do that in the main thread and spawn new threads for processing the data
2. Using a [profiler that understands disk access](http://www.linuxprogrammingblog.com/io-profiling) would give more concrete information on whether we have a IO-bottleneck

**To test:**

All of the existing tests were unchanged. Reviewing the code is all that is needed.

*There is no associated issue.*

*This does not require release notes* because this is an internal code refactor.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by peterfpeterson** [comment]
Using `iotop` suggests that the majority of the time (in cold cache case) is spent waiting on file io.

File: Framework/DataHandling/src/ProcessBankData.cpp [hunk 1]
```diff
@@ -33,6 +33,33 @@ ProcessBankData::ProcessBankData(
   m_cost = static_cast<double>(numEvents);
 }
 
+namespace {
+// this assumes that last_pulse_index is already to the point of including this
+// one so we only need to search forward
+inline size_t
+getPulseIndex(const size_t event_index, const size_t last_pulse_index,
+              const boost::shared_ptr<std::vector<uint64_t>> &event_index_vec) {
```
Comment (quantumsteve) [inline Framework/DataHandling/src/ProcessBankData.cpp]: consider just passing `const std::vector<uint64_t> &event_index_vec`

Comment (peterfpeterson) [reply Framework/DataHandling/src/ProcessBankData.cpp]: Is that faster?

Comment (quantumsteve) [reply Framework/DataHandling/src/ProcessBankData.cpp]: It makes the interface cleaner. I pointer may be NULL but a reference cannot be.

In principle, you should have a check for NULL in this function, while passing a reference moves that responsibility to the calling code.

**Comment by peterfpeterson** [comment]
I deleted the last two commits which attempted to reuse the file handle and either gave lots of warnings on ubuntu 18.04 (penultimate) or broke the unit test (ultimate). This is now ready to be reviewed again and merged.

**Review by AndreiSavici — APPROVED** [review]

**Pull request merged**